cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/system/SystemWorkerTwillRunnable.java

```diff
@@ -81,26 +114,75 @@
   private LogAppenderInitializer logAppenderInitializer;
   private MetricsCollectionService metricsCollectionService;
 
-  public SystemWorkerTwillRunnable(String cConfFileName, String hConfFileName) {
-    super(ImmutableMap.of("cConf", cConfFileName, "hConf", hConfFileName));
+  public SystemWorkerTwillRunnable(String cConfFileName, String hConfFileName, String sConfFileName) {
+    super(ImmutableMap.of("cConf", cConfFileName, "hConf", hConfFileName, "sConf", sConfFileName));
   }
 
   @VisibleForTesting
-  static Injector createInjector(CConfiguration cConf, Configuration hConf) {
+  static Injector createInjector(CConfiguration cConf, Configuration hConf, SConfiguration sConf) {
     List<Module> modules = new ArrayList<>();
 
-    CoreSecurityModule coreSecurityModule = CoreSecurityRuntimeModule.getDistributedModule(cConf);
+    CoreSecurityModule coreSecurityModule = new FileBasedCoreSecurityModule() {
+      @Override
+      protected void bindKeyManager(Binder binder) {
+        binder.bind(KeyManager.class).to(FileBasedKeyManager.class).in(Scopes.SINGLETON);
+        expose(KeyManager.class);
+      }
+    };
 
-    modules.add(new ConfigModule(cConf, hConf));
+    modules.add(new ConfigModule(cConf, hConf, sConf));
     modules.add(RemoteAuthenticatorModules.getDefaultModule());
     modules.add(new LocalLocationModule());
     modules.add(new IOModule());
-    modules.add(new AuthenticationContextModules().getMasterWorkerModule());
+    modules.add(new AuthenticationContextModules().getMasterModule());
     modules.add(coreSecurityModule);
     modules.add(new MessagingClientModule());
     modules.add(new SystemAppModule());
     modules.add(new MetricsClientRuntimeModule().getDistributedModules());
 
+    modules.addAll(Arrays.asList(
+      // Always use local table implementations, which use LevelDB.
+      // In K8s, there won't be HBase and the cdap-site should be set to use SQL store for StructuredTable.
+      new DataSetServiceModules().getStandaloneModules(),
+      // The Dataset set modules are only needed to satisfy dependency injection
+      new DataSetsModules().getStandaloneModules(),
+      new MetricsStoreModule(),
+      new MessagingClientModule(),
+      new ExploreClientModule(),
+      new AuditModule(),
+      new AuthorizationModule(),
+      new AuthorizationEnforcementModule().getMasterModule(),
+      Modules.override(new AppFabricServiceRuntimeModule(cConf).getDistributedModules())
+        .with(new AbstractModule() {
+          @Override
+          protected void configure() {
+            bind(StorageProviderNamespaceAdmin.class).to(
+              LocalStorageProviderNamespaceAdmin.class);
+          }
+        }),
+      new ProgramRunnerRuntimeModule().getDistributedModules(true),
+      new MonitorHandlerModule(false),
+      new SecureStoreServerModule(),
+      new OperationalStatsModule(),
```

Comment:
is this needed?

Reply:
We'll test this to make sure it's necessary once PR1 & PR2 are merged.